client/packages/android/app/src/main/java/org/openmsupply/client/CertWebViewClient.java

```diff
@@ -65,59 +69,105 @@ private X509Certificate get_x509(SslCertificate certificate) {
     }
 
-    public CertWebViewClient(Bridge bridge, File filesDir) {
+    public CertWebViewClient(Bridge bridge, File filesDir, NativeApi nativeApi) {
         super(bridge);
 
+        savedCertFingerprints = bridge.getContext().getSharedPreferences("savedCertFingerprints", Context.MODE_PRIVATE);
+        this.nativeApi = nativeApi;
         this.filesDir = filesDir;
     }
 
-    @SuppressLint("WebViewClientOnReceivedSslError")
-    @Override
-    public void onReceivedSslError(WebView view, SslErrorHandler handler,
-                                   SslError error) {
-        // TODO store and trust previous cert
-        if (!error.getUrl().contains("localhost")) {
-            handler.proceed();
-            return;
-        }
+    private boolean validateLocalCertificate(SslCertificate targetCert) {
         // If there is a ssl error, check if the request was trying to reach our local trusted
         // remote server. For this:
         // 1) load self signed remote server certificate from local storage
         // 2) validate that the request was trying to reach the server by validating that the
         //    target certificate from the request is matching our known self signed certificate
         Certificate selfSignedCert = this.get_self_signed_cert();
-        SslCertificate targetCert = error.getCertificate();
         if (selfSignedCert == null || targetCert == null) {
-            super.onReceivedSslError(view, handler, error);
-            return;
+            return false;
         }
         Certificate targetX509Cert = get_x509(targetCert);
         if (targetX509Cert == null) {
             Log.e(TAG, "Failed to extract x509 request target certificate");
-            super.onReceivedSslError(view, handler, error);
-            return;
+            return false;
         }
 
         try {
             targetX509Cert.verify(selfSignedCert.getPublicKey());
-            handler.proceed();
-            return;
+            return true;
         } catch (Exception e) {
             Log.e(TAG, "Invalid request target certificate" + e);
         }
 
-        super.onReceivedSslError(view, handler, error);
+        return false;
+    }
+
+    private boolean validateNonLocalCertificate(SslCertificate targetCert, NativeApi.omSupplyServer connectedServer) {
+        // Calculate SSL fingerprint
+        PublicKey publicKey = get_x509(targetCert).getPublicKey();
+        MessageDigest md = null;
+        try {
+           md = MessageDigest.getInstance("SHA-256");
+        } catch (Exception e) {
+            Log.e(TAG, "SHA-256 algorithm is missing" + e);
+            return false;
+        }
+        md.update(publicKey.getEncoded());
+        String fingerprint = Base64.encodeToString(md.digest(), Base64.DEFAULT).trim();
+
+        // Match SSL fingerprint for server stored in app data
+        // Match by hardware id and port
+        String identifier = connectedServer.getHardwareId() + "-" + connectedServer.getPort();
```

Comment:
Why is hardware id and port important? As long as the key matches we should be fine doing the https connection, which is the only thing we are interested right here.

Reply:
A client could connect to multiple servers, there is UI for this (can go back to discovery from login or initialisation page). Use case for this is consultant having client installed and going to multiple sites, they can connect to local remote servers.

This is very similar to how ssh client works, except we don't store by domain/ip but by hardware id and port since local ip can change and can have multiple server on the same ip (at least for dev)